actors/builtin/miner/expiration_queue.go

```diff
@@ -698,3 +716,126 @@ func groupSectorsByExpiration(sectorSize abi.SectorSize, sectors []*SectorOnChai
 	})
 	return sectorEpochSets
 }
+
+// Groups sectors into sets based on their Expiration field.
+// If sectors are not found in the expiration set corresponding to their expiration field
+// (i.e. they have been rescheduled) traverse expiration sets to for groups where these
+// sectors actually expire.
+// Groups will be returned in expiration order, earliest first.
+func (q *ExpirationQueue) findSectorsByExpiration(sectorSize abi.SectorSize, sectors []*SectorOnChainInfo) ([]sectorExpirationSet, error) {
+	declaredExpirations := make(map[abi.ChainEpoch]bool, len(sectors))
+	sectorsByNumber := make(map[uint64]*SectorOnChainInfo, len(sectors))
+	emptyBitfield := bitfield.New()
+	allNotFound := &emptyBitfield
+	expirationGroups := make([]sectorExpirationSet, 0, len(declaredExpirations))
+
+	for _, sector := range sectors {
+		qExpiration := q.quant.QuantizeUp(sector.Expiration)
+		declaredExpirations[qExpiration] = true
+		allNotFound.Set(uint64(sector.SectorNumber))
+		sectorsByNumber[uint64(sector.SectorNumber)] = sector
+	}
+
+	// Traverse expiration sets first by expected expirations. This will find all groups if no sectors have been rescheduled.
+	// This map iteration is non-deterministic but safe because we sort by epoch below.
+	for expiration := range declaredExpirations { //nolint:nomaprange // result is subsequently sorted
+		es, err := q.mayGet(expiration)
+		if err != nil {
+			return nil, err
+		}
+
+		// create group from overlap
+		var group sectorExpirationSet
+		group, allNotFound, err = groupExpirationSet(sectorSize, sectorsByNumber, allNotFound, es, expiration)
+		if err != nil {
+			return nil, err
+		}
+		if len(group.sectors) > 0 {
+			expirationGroups = append(expirationGroups, group)
+		}
+	}
+
+	// If sectors remain, traverse next in epoch order. Remaining sectors should be rescheduled to expire soon, so
+	// this traversal should exit early.
+	if empty, err := allNotFound.IsEmpty(); err != nil {
+		return nil, err
+	} else if !empty {
+		err := q.traverse(func(epoch abi.ChainEpoch, es *ExpirationSet) (bool, error) {
+			// skip if we've processed this epoch above
```

Comment:
Please expand the comment, noting that if a sector was in this epoch due to rescheduling it would have already been included in the first loop.